### PR TITLE
batches: add docs and changelog for publish from UI

### LINKED ISSUE #18277: 🚀 Publish batch changes from GUI
## Background

We want Batch Changes to be widely adopted within a company. To achieve that, we need to make sure that the effort required to onboard Batch Changes should be low. 

<p></p>

Companies with competing internal tooling typically have increasing internal adoption as a success criteria, and  developer experience and UI coverage are a key driver.

## Problem

Publishing changesets requires switching context from CLI to the GUI, then back to the CLI. This is creating friction for new users to onboard, and a bad ongoing experience for active users.

Specifically,  
When I create a campaign, I start by defining a campaign spec. As it is a new campaign, I set the campaign spec attribute `published` to `false`, indicating that all the changesets created by the campaigns will be `unpublished`.  
I run `src campaigns preview -f my.campaign.yml`. 

<p></p>

The command opens a link that takes me to the UI to preview my campaign. I choose to *apply* it in the GUI (create it on the Sourcegraph instance). The changes look good.  
Now I want to `publish` the changesets (open changesets on the code host).

To do that, I need to go back to the spec (locally), change `published` to `true`, then run `src campaigns apply -f my.campaign.yml`.

create spec (CLI) -> preview (GUI) -> apply (GUI) -> publish (**CLI**) -> apply (GUI).

<p></p>

**Customer feedback**

See [productboard](https://sourcegraph.productboard.com/roadmap/2263724-campaigns-releases/features/6956091/insights).

## Proposed solution

**Success criteria**

Success means

- Users can run `src batch preview -f`, then never leave the GUI

- Users that run Batch Changes in CI or automation can still use the current spec-based workflow (no breaking change)

- The specs is the primary source of truth

<p></p>

**Solution**

Chris sketched out a solution in his [offsite talk](https://docs.google.com/document/d/1Sq4yy-eqiwvasSIIqgb_2bcSIJftCyZcUfbyHsJxKYw/edit), that served as the starting point for this. It allows for 2 workflows, the spec & CLI workflow, and a GUI workflow.

<p></p>

Published is optional in the specs.

<p></p>

If there is no published in the specs, the source of truth is the GUI. A newly created batch change with no published statement in the spec is not published. 

- I can go through the GUI to publish/ un-publish them

- and individually pick what changesets to publish/ un-publish

<p></p>

If a user then adds a published statement to the spec and re-applies it, then the status that exists in the GUI (that is, which changesets are published or not) is overridden by what is in the spec. In other words, the spec takes over as the source of truth.

<p></p>

If a user applies a spec with a publish statement, then removes the publish statement and re-applies, the GUI becomes the source of truth.

<p></p>

Since this change introduces two alternative flows to publish changesets, it needs to be clear for users why publishing though the GUI is available or not. For example, when the spec has a `published` statement, I can see the `published` button in the GUI, but it's greyed out with a hover "This batch change spec contains a `published` statement, so publishing is controlled by the spec".

<p></p>

**Related issues in scope**

- <https://github.com/sourcegraph/sourcegraph/issues/17541>

<p></p>

## **Discussion**

*We can use this as an experiment to driver future decisions on the behaviour of published.*

<p></p>

That change eliminates one context switch, but there is still unnecessary context switching. Ideally, an entire campaign lifecycle should be possible from a single interface (GUI / CLI)

## **Release process**

We are not feature flagging this feature since it does not break any existing flow.

### PR DESCRIPTION
**Pull request by LawnGnome**
This is the final part eight, and therefore closes #18277.

Nothing very interesting here: I haven't changed the defaults that we recommend here, since we talked about holding off until we have more functionality on the batch spec preview page.

### CODE REVIEW

File: doc/batch_changes/how-tos/bulk_operations_on_changesets.md [hunk 1]
```diff
@@ -25,6 +25,7 @@ Bulk operations allow a single action to be performed across many changesets in
...
 - <span class="badge badge-experimental">Experimental</span> Merge: Only available if filtering by state `open`. Tries to merge the selected changesets on the code hosts. Due to the nature of changesets, there are many states in which a changeset is not mergeable. This won't break the entire bulk operation, but single changesets may not be merged after the run for this reason. The bulk operations tab lists those where merging failed below the bulk operation in that case. In the confirmation modal, you can select to merge using the squash merge strategy. This is supported on both GitHub and GitLab, but not on Bitbucket Server. In this case, regular merges are always used for merging the changesets.
 - Close: Only available if filtering by state `open` or `draft`. Tries to close the selected changesets on the code hosts.
+- Publish: Publishes the selected changesets, provided they don't have a [`published` field](../references/batch_spec_yaml_reference.md#changesettemplate-published) in the batch spec. You can choose between draft and normal changesets in the confirmation modal.
 
 _More types coming soon._
```
Comment (courier-new) [inline doc/batch_changes/how-tos/bulk_operations_on_changesets.md:30]: Not strictly relevant to this PR but I feel like we've got a good set of supported types here now and can probably remove this message.

File: doc/batch_changes/how-tos/publishing_changesets.md [hunk 2]
```diff
@@ -19,6 +33,75 @@ For more information, see "[Code host interactions in Batch Changes](../explanat
 
 ## Publishing changesets
 
+You can publish changesets either by [setting the `published` field in the batch spec](#within-the-spec), or [through the Sourcegraph UI](#within-the-ui). Both workflows are described in full below.
+
+A brief summary of the pros and cons of each workflow is:
```
Comment (courier-new) [inline doc/batch_changes/how-tos/publishing_changesets.md:38]: Really love this summary table!!

File: doc/batch_changes/quickstart.md [hunk 3]
```diff
@@ -132,6 +132,8 @@ Now that you have credentials set up, you can publish the changesets in the batc
 
     > NOTE: You can also create or update a batch change by running `src batch apply`. This skips the preview stage, and is especially useful when updating an existing batch change.
 
+> NOTE: You can also publish directly from Sourcegraph by omitting the `published` field from your batch spec. This is described in more detail in "[Publishing changesets to the code host](how-tos/publishing_changesets.md#publishing-changesets)".
+
```
Comment (courier-new) [inline doc/batch_changes/quickstart.md:136]: Is this new note intentionally indented differently from the previous one?

**Review by courier-new — APPROVED** [review]
Seems well-written to me, as always!! I think you're gone by now so I'm comfortable if we want to merge this, just left a few nits.

**Review by eseliger — APPROVED** [review]
Nothing to add to Kellis notes, LGTM!

File: doc/batch_changes/quickstart.md (hunk 3 above)
Comment (LawnGnome) [reply doc/batch_changes/quickstart.md:136]: Yes, the previous note is part of the last list item, whereas this note is back at the top level.

File: doc/batch_changes/how-tos/bulk_operations_on_changesets.md (hunk 1 above)
Comment (LawnGnome) [reply doc/batch_changes/how-tos/bulk_operations_on_changesets.md:30]: Agreed. I'll take it out in this PR.

**Pull request merged**